### ISSUE TITLE
mojom bindings: should generate functions to check that a value is a known valid value for each mojom enum

### ISSUE THREAD

**Issue by viettrungluu-cr** [domokit/mojo#332]
This is a baby step towards stricter/automatic validation of enums. I'm not sure what exact form such a function should take. E.g., in C++:
```bool IsKnownValidMyEnumValue(MyEnum value);```
or maybe
```bool IsKnownValidMyEnumValue(int32_t value);```
(not claiming that `value` is actually a valid `MyEnum`), or maybe
```bool IsKnownValidValue(MyEnum value);```
(relying on overloading), or maybe it should be `IsKnownValidMyEnum` (the "value" part being redundant).

Presumably, it'd be nice to have such functions (or similar, as appropriate) for each language our bindings support.

**Comment by viettrungluu-cr** [comment]
@vardhan-cr Do you want to take a look at this, maybe? (You can start with C++, say. It's not super-urgent to have this for all languages.)

**Comment by j9brown** [comment]
Might want to consider the impact on code size if these functions are not
used often.

**Comment by vardhan-cr** [comment]
sure, I'll take a look.

What do you think about making this a templated function that enums can specialize?
i.e.,  given:

template <typename E>
bool IsValidEnumValue<E>(int32_t);

generated enums can implement a specialized version:
template <>
bool IsValidEnumValue<MyEnum>(int32_t) { .. }

j9brown@,
this feature is based on an email thread that you may not have seen.  (fwd'ing it to you).  To quote:

"The general reason for this is that enums may be extensible (so we can't tell if a given value is bad, or from an unknown/future version)."

i hadn't considered code size would be a problem -- however, if enums are "well structured" (ie integrals representing the enum values are continuous), it'll just be a single if statement of "if (enum_value >= 0 && enum_value < NUM_VALUES) return true".  maybe in the future, we can consider a way to disable/using validation support when compiling mojom, if that makes sense, to reserve code space?

**Comment by viettrungluu-cr** [comment]
@j9brown I don't think it should affect code size should be a concern -- these things are statically linked into their targets, so their implementations would be dropped by the linker if they're not used.

**Comment by viettrungluu-cr** [comment]
@j9brown I don't think it should affect code size should be a concern -- these things are statically linked into their targets, so their implementations would be dropped by the linker if they're not used.

**Comment by viettrungluu-cr** [comment]
@vardhan-cr I don't have a particularly strong opinion. You may want to solicit opinions on mojo-dev. One thing to consider, though: If MyEnum is in the foo.bar "module" (thus in the foo::bar) namespace, then the non-templatized versions would naturally be put in the foo::bar namespace as well, whereas the templatized function would be in some standard namespace (and depending on what namespace it's used in, you'd have to qualify the template parameter appropriately. Not sure if this is good or bad.
